pkg/operator/controller/dns/controller.go

```diff
@@ -0,0 +1,178 @@
+	"k8s.io/client-go/tools/record"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	configv1 "github.com/openshift/api/config/v1"
+	operatorv1 "github.com/openshift/api/operator/v1"
+
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	runtimecontroller "sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	controllerName = "dns_controller"
+)
+
+var log = logf.Logger.WithName(controllerName)
+
+func New(mgr manager.Manager, operatorNamespace string, dnsProvider dns.Provider) (runtimecontroller.Controller, error) {
+	reconciler := &reconciler{
+		client:            mgr.GetClient(),
+		cache:             mgr.GetCache(),
+		recorder:          mgr.GetEventRecorderFor(controllerName),
+		operatorNamespace: operatorNamespace,
+		dnsProvider:       dnsProvider,
+	}
+	c, err := runtimecontroller.New(controllerName, mgr, runtimecontroller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+	if err := c.Watch(&source.Kind{Type: &iov1.DNSRecord{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+type reconciler struct {
+	client            client.Client
+	cache             cache.Cache
+	recorder          record.EventRecorder
+	operatorNamespace string
+	dnsProvider       dns.Provider
+	dnsConfig         *configv1.DNS
+}
+
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	// TODO: If the new revision has changed, undefined stuff happens (resource leaks?)
+	dnsConfig := &configv1.DNS{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: "cluster"}, dnsConfig); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to get dns 'cluster': %v", err)
+	}
+
+	record := &iov1.DNSRecord{}
+	if err := r.client.Get(context.TODO(), request.NamespacedName, record); err != nil {
+		if errors.IsNotFound(err) {
+			log.Info("dnsrecord not found; reconciliation will be skipped", "request", request)
+			return reconcile.Result{}, nil
+		} else {
+			log.Error(err, "failed to get dnsrecord; will retry", "dnsrecord", request.NamespacedName)
+			return reconcile.Result{RequeueAfter: 15 * time.Second}, nil
+		}
+	}
+
+	// If the DNS record was deleted, clean up and return.
+	if record.DeletionTimestamp != nil {
+		if err := r.delete(record); err != nil {
+			log.Error(err, "failed to delete dnsrecord; will retry", "dnsrecord", record)
+			return reconcile.Result{RequeueAfter: 15 * time.Second}, nil
+		}
+		return reconcile.Result{}, nil
+	}
+
+	ingressName, ok := record.Labels[manifests.OwningIngressControllerLabel]
+	if !ok {
+		return reconcile.Result{}, fmt.Errorf("missing owner label")
```

Comment:
Is it useful to requeue in this case?

Reply:
Added some logging and infrequent retries here just in case? Not clear cut what the best behavior is around invalid states in general (should we surface those in a way that's more easily analyzed? Like, in this case we're literally seeing corruption/invalid states, and it seems like we'd want to do more than just log and wait for the problem to fix itself).

Seems like a systemic problem we need to address